lilypad/_opentelemetry/_opentelemetry_outlines/__init__.py

```diff
@@ -0,0 +1,180 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Modifications copyright (C) 2024 Mirascope
+from __future__ import annotations
+
+from collections.abc import Collection
+from contextlib import suppress
+from typing import Any
+
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.utils import unwrap
+from opentelemetry.semconv.schemas import Schemas
+from opentelemetry.trace import get_tracer
+from wrapt import wrap_function_wrapper
+
+from lilypad._opentelemetry._opentelemetry_outlines.patch import (
+    model_generate,
+    model_generate_async,
+    model_generate_stream,
+)
+
+_patched_targets: list[tuple[str, str, str]] = []
+
+
+class OutlinesInstrumentor(BaseInstrumentor):
+    def instrumentation_dependencies(self) -> Collection[str]:
+        return ("outlines>=0.1.10,<1.0",)
+
+    def _instrument(self, **kwargs: Any) -> None:
+        """Enable Outlines instrumentation."""
+        tracer_provider = kwargs.get("tracer_provider")
+        tracer = get_tracer(
+            __name__,
+            "0.1.0",  # Lilypad version or plugin version
+            tracer_provider,
+            schema_url=Schemas.V1_28_0.value,
+        )
+        # Instrument ExLlamaV2Model
+        wrap_function_wrapper(
+            "outlines.models.exllamav2",
+            "ExLlamaV2Model.generate",
+            model_generate(tracer),
+        )
+        _patched_targets.append(
+            ("outlines.models.exllamav2", "ExLlamaV2Model", "generate")
+        )
```

Comment:
@willbakst @brenkao 
While users can directly call model.generate, they can also use it like `outlines.generate.json(model, obj)(prompt)` or `outlines.generate.text(model)(prompt)`.
Since all these patterns ultimately call the generate method (or similar methods) defined in the model class as shown above, we're covering all call patterns. However, I'd like to confirm: do we need to patch methods like `json()` and `text()` as well, or is it sufficient since they eventually call the patched method anyway?

Reply:
if generate gets called eventually, then we shouldnt need to patch other methods. we just need to be able to handle json and text data.